### PR TITLE
QML chat implementation

### PR DESCRIPTION
**Pull request by valentynfaychuk**
* Changed general chat interfaces
* Added qml chat support
* Changed dock behavior & added multiple themes support
* Added new special chat icons
* Embedded QML scene into application

### CODE REVIEW

File: src/chatwindow/chatdisabled.qml [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            Control {
+                anchors.horizontalCenter: parent.horizontalCenter
+                //anchors.bottom: unsatisfiedUserImage.top
+                width: warningText.implicitWidth + 2
+
+                height: 30
+                Label {
+                    id: warningText
+                    text: qsTr("Log in to use chat")
+                    color: {
```
Comment (anantonyk) [inline src/chatwindow/chatdisabled.qml]: You use code below twice, so it maybe better to make a special function for it

File: src/chatwindow/chatelements/AvailUsers.qml [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+    DelegateModel {
+        id: availableUsersModel
+
+        model: UsersModel {
+            list: usersList
+
+            onDataChanged: {
+                /*
+                console.log("QML DATA HAS CHANGED");
```
Comment (anantonyk) [inline src/chatwindow/chatelements/AvailUsers.qml]: remove commented lines

File: src/chatwindow/chatwidget.h [hunk 3]
```diff
@@ -1,48 +1,61 @@
...
-    QStringList  mOnlineUsers;
-    QStringList  mConnectedUsers;
-    QString      mUserName;
+public:
 
-    QPlainTextEdit *mpFeed;
-    QLineEdit      *mpEnterLine;
+    ChatWidget();
+    ChatWidget(ChatWidget const&)             = delete;
+    ChatWidget & operator=(ChatWidget const&) = delete;
```
Comment (anantonyk) [inline src/chatwindow/chatwidget.h]: ChatWidget& operator=(ChatWidget const&) = delete;
and the same below

File: src/chatwindow/chatwidget.h [hunk 4]
```diff
@@ -1,48 +1,61 @@
...
+
+    // Name of the current user
+    QString      mUserName;
+
+    // Lists of online as well as connected users
+    QStringList  mOnlineUsers;
+    QStringList  mConnectedUsers;
+
+    // Users list, editor & message input fields
+    QListWidget    * mpUsersList;
```
Comment (anantonyk) [inline src/chatwindow/chatwidget.h]: QListWidget    *mpUsersList;
and the same below

File: src/chatwindow/chatwindowdock.cpp [hunk 5]
```diff
@@ -1,69 +1,100 @@
...
-}
+    connect(mpChatWidget, &ChatWidgetInterface::stopSharingRequested,
+            this,         &ChatWindowDock::sendRequestOnStopSharing,
+            Qt::UniqueConnection);
+
+    connect(mpChatWidget, &ChatWidgetInterface::messageSent,
+            this,         &ChatWindowDock::shareMessageOnSend,
+            Qt::UniqueConnection);
 
+    setWindowTitle("Chat");
```
Comment (anantonyk) [inline src/chatwindow/chatwindowdock.cpp]: I recommend you to set a variable, for example const char[] windowTitle = "Chat";

File: src/chatwindow/chatwidgetinterface.h [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+// ==========================================================================================
+//                                                                             CHAT INTERFACE
+// ==========================================================================================
+class ChatWidgetInterface : public QWidget
+{
+    Q_OBJECT
+
+protected:
+
+    explicit ChatWidgetInterface(QWidget *qObject = nullptr) : QWidget(qObject) {}
```
Comment (anantonyk) [inline src/chatwindow/chatwidgetinterface.h:15]: ... = nullptr): QWidget ...

File: src/chatwindow/qmlchatwidget.cpp [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+}
+// ==========================================================================================
+// ==========================================================================================
+// ==========================================================================================
+//                                                             APPEND NEW MESSAGE TO THE CHAT
+void QmlChatWidget::appendMessage(const QString &messageAuthor,
+                                  const QString &messageBody)
+{
+    // Here we IGNORE messageAuthor because it is solely NEEDED FOR BACK COMPATIBILITY
+    if(!mpMessagesController)
```
Comment (anantonyk) [inline src/chatwindow/qmlchatwidget.cpp]: if (!mpMessagesController)

File: src/chatwindow/usersmodel/chatusersmodel.cpp [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+
+    // Switch upon attributes to be changed
+    switch(role)
+    {
+    case UserNameRole:
+        // Idle
+        break;
+    case UserConnectedRole:
+        // User can change only "isConnected" attribute
+        if (newValue.toBool())
```
Comment (anantonyk) [inline src/chatwindow/usersmodel/chatusersmodel.cpp:85]: embed this if statement into function

File: src/network/defaultlocalconnector.cpp [hunk 9]
```diff
@@ -106,7 +106,7 @@ void DefaultLocalConnector::clearOutdatedAttributesOnTimerTick()
     if (!outdatedServerNames.empty())
     {
         bool isAnyConnectedAttributesOutdated = false;
-        for(const auto & serverName : outdatedServerNames)
+        for(const auto &serverName : outdatedServerNames)
```
Comment (anantonyk) [inline src/network/defaultlocalconnector.cpp]: for (const auto &serverName : outdatedServerNames)

File: src/network/defaultlocalconnector.cpp [hunk 10]
```diff
@@ -106,7 +106,7 @@ void DefaultLocalConnector::clearOutdatedAttributesOnTimerTick()
...
         {
             if(!popFromConnectedAttributes(serverName).empty())
```
Comment (anantonyk) [inline src/network/defaultlocalconnector.cpp]: if (!popFromConnectedAttributes(serverName).empty())

Comment (anantonyk) [inline src/network/defaultlocalconnector.cpp]: if (!popFromConnectedAttributes(serverName).empty())

File: src/chatwindow/chatwidgetinterface.h [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+    // Possible chat themes & current theme variable
+    enum Theme {
```
Comment (anantonyk) [inline src/chatwindow/chatwidgetinterface.h]: use strongly typed enum

File: src/chatwindow/chatelements/MsgsLifeline.qml [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+import QtQml.Models 2.12
+import "basicelements" as BasicElements
+import PairStormChat 1.0
+
+Item {
+    id: messagesHistory
+
+    Rectangle {
+        anchors.fill: parent
+        color: "#5a8db8"
```
Comment (anantonyk) [inline src/chatwindow/chatelements/MsgsLifeline.qml]: write comment with color name

File: src/network/defaultlocalconnector.cpp (hunk 10 above)
Comment (valentynfaychuk) [reply src/network/defaultlocalconnector.cpp]: done

File: src/chatwindow/qmlchatwidget.cpp (hunk 7 above)
Comment (valentynfaychuk) [reply src/chatwindow/qmlchatwidget.cpp]: done

File: src/network/defaultlocalconnector.cpp (hunk 9 above)
Comment (valentynfaychuk) [reply src/network/defaultlocalconnector.cpp]: fixed

File: src/chatwindow/chatwidgetinterface.h (hunk 6 above)
Comment (valentynfaychuk) [reply src/chatwindow/chatwidgetinterface.h:15]: No matter: see code conventions

**Review by PeterLysyk — APPROVED** [review]
Please, create smaller pull requests. It's so diffucult to review almost 4k lines of code

**Review by anantonyk — APPROVED** [review]

File: src/chatwindow/chatdisabled.qml (hunk 1 above)
Comment (valentynfaychuk) [reply src/chatwindow/chatdisabled.qml]: Now these actions are preformed inside JS functions

File: src/chatwindow/chatwindowdock.cpp (hunk 5 above)
Comment (valentynfaychuk) [reply src/chatwindow/chatwindowdock.cpp]: Done

File: src/chatwindow/chatelements/AvailUsers.qml (hunk 2 above)
Comment (valentynfaychuk) [reply src/chatwindow/chatelements/AvailUsers.qml]: All unnecessary comments now removed

File: src/chatwindow/chatwidget.h (hunk 3 above)
Comment (valentynfaychuk) [reply src/chatwindow/chatwidget.h]: Done

File: src/chatwindow/chatwidget.h (hunk 4 above)
Comment (valentynfaychuk) [reply src/chatwindow/chatwidget.h]: Fixed

**Pull request merged**